### PR TITLE
Refactoring plot_proc.py for use in fusrr

### PR DESCRIPTION
**Pull request by clmould**
## Description

Refactoring `plot_proc.py` for use in fusrr. Parts of `plot_proc` were copied and pasted into `fusrr` so refactoring these parts into separate files to be accessed by both `plot_proc` and `fusrr` without anything being copied and pasted between the two. The parts being refactored here are functions for the plasma, tf coils, pf coils, cryostat, blanket, shield and vacuum vessel.
Also refactored function for the first wall and added first wall plotting into `fusrr`.
## Checklist

I confirm that I have completed the following checks:

- [x] I have justified any large differences in the regression tests caused by this pull request in the comments.
- [x] I have added new tests where appropriate for the changes I have made.
- [x] If I have had to change any existing unit or integration tests, I have justified this change in the pull request comments.
- [x] If I have made documentation changes, I have checked they render correctly.
- [x] I have added documentation for my change, if appropriate.

### CODE REVIEW

**Comment by oliverfunk** [comment]
From my end, this LGTM!

File: process/geometry/tfcoil_geometry.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+            center_x=x4,
+            center_z=y4 - tfcth,
+            width=tfcth,
+            height=(y2 - y4 + 2.0 * tfcth),
+        )
+    )
+    return_rects.append(
+        RectangleGeometry(center_x=x5, center_z=y5 - tfcth, width=x4 - x5, height=tfcth)
+    )
+    return_rects.append(
```
Comment (timothy-nunn) [inline process/geometry/tfcoil_geometry.py:63]: Again, what are all of these rectangles? A comment would be helpful to see what this function is doing

File: process/geometry/tfcoil_geometry.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    :param y5: vertical location of arc point 5
+    :type y5: float
+    :param tfcth: inboard tf coil thickness
+    :type tfcth: float
+    :param rtangle: angle used in tf coil parametrization
+    :type rtangle: float
+    :param rtangle2: angle used in tf coil parametrization
+    :type rtangle2: float
+    :return: radial and vertical coordinates for tf coils
+    :rtype: Tuple[List[RectangleGeometry], list]
```
Comment (timothy-nunn) [inline process/geometry/tfcoil_geometry.py:111]: `:rtype: Tuple[List[RectangleGeometry], List[float]]`? Double check that but I think thats a more descriptive type

Comment (timothy-nunn) [reply process/geometry/tfcoil_geometry.py:111]: Wait no maybe its a List[List[float]] looking at the code. Need a better description of the types and meanings of the variables involved

File: process/geometry/tfcoil_geometry.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+        ang1=0,
+        ang2=-rtangle,
+    )
+    # Vertical leg
+    return_rects.append(
+        RectangleGeometry(
+            center_x=x5 - tfcth, center_z=y5, width=tfcth, height=(y1 - y5)
+        )
+    )
+    return_verts = [verts1, verts2, verts3, verts4]
```
Comment (timothy-nunn) [inline process/geometry/tfcoil_geometry.py:189]: Again, what is this?

File: process/geometry/utils.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
+"""
+Module to hold plotting functions, used in plot_proc.py, which are common to multiple reactor components
+"""
+import numpy as np
+
+
+def plotdh(r0: float, a: float, triang: float, kap: float) -> tuple:
+    """Plots half a thin D-section, centred on z = 0
```
Comment (timothy-nunn) [inline process/geometry/utils.py:8]: Does this *plot* the D or does it just return the points to be plot? I cant see any plotting logic

File: process/geometry/utils.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    :param r0: major radius of centre
+    :type r0: float
+    :param a: horizontal radius
+    :type a: float
+    :param triang: plasma triangularity
+    :type triang: float
+    :param kap: plasma elongation
+    :type kap: float
+    :return: radial and vertical coordinates of D-section
+    :rtype: tuple
```
Comment (timothy-nunn) [inline process/geometry/utils.py:19]: `Tuple[List[float], List[float]]`?

File: process/geometry/utils.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+    :param outthk: outer thickness
+    :type outthk: float
+    :param toppt: top point
+    :type toppt: float
+    :param topthk: top thickness
+    :type topthk: float
+    :param triang: plasma triangularity
+    :type triang: float
+    :return: radial and vertical coordinates of D-section
+    :rtype: tuple
```
Comment (timothy-nunn) [inline process/geometry/utils.py:53]: Tuple of. Type and description

File: process/geometry/utils.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+    a1: float = 0,
+    a2: float = 0,
+    b1: float = 0,
+    b2: float = 0,
+    x0: float = 0,
+    y0: float = 0,
+    ang1: float = 0,
+    ang2: float = np.pi / 2,
+) -> list:
+    """Fills the space between two concentric ellipse sectors
```
Comment (timothy-nunn) [inline process/geometry/utils.py:86]: I think it more returns the vertices of a shape which, when filled... "fills the space between two concentric ellipse sectors"

File: process/geometry/vacuum_vessel_geometry.py [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+Calculate radial and vertical coordinates for the geometry of the vacuum vessel
+"""
+from typing import List
+from dataclasses import dataclass
+import numpy as np
+from process.geometry.utils import plotdh
+
+
+@dataclass
+class VacuumVesselGeometry:
```
Comment (timothy-nunn) [inline process/geometry/vacuum_vessel_geometry.py:11]: Same question about ArbitraryGeometry types

File: process/geometry/vacuum_vessel_geometry.py [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+    :param radx_inner: outboard radius of inner surface of vacuum vessel
+    :type radx_inner: float
+    :param rminx_inner: inboard radius of inner surface of vacuum vessel
+    :type rminx_inner: float
+    :param cumulative_lower: cumulative vertical thicknesses of components below the midplane
+    :type cumulative_lower: dict
+    :param lower: vertical thicknesses of components below the midplane
+    :type lower: dict
+    :return: VacuumVesselGeometry - dataclass returning vacuum vessel radial and vertical coordinates
+    :rtype: DataClass
```
Comment (timothy-nunn) [inline process/geometry/vacuum_vessel_geometry.py:52]: Same comment as above re type of return

File: process/geometry/vacuum_vessel_geometry.py [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+        radx_inner=radx_inner,
+        radx_outer=radx_outer,
+        rminx_inner=rminx_inner,
+        rminx_outer=rminx_outer,
+        triang=triang,
+    )
+
+    rs = np.concatenate([rs_lower_2, rs_lower_1[::-1], rs_upper_1, rs_upper_2[::-1]])
+    zs = np.concatenate([zs_lower_2, zs_lower_1[::-1], zs_upper_1, zs_upper_2[::-1]])
+    return VacuumVesselGeometry(
```
Comment (timothy-nunn) [inline process/geometry/vacuum_vessel_geometry.py:76]: Same comment that these types won't match the type hint

File: process/geometry/vacuum_vessel_geometry.py [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+    :param radx_outer: outboard radius of outer surface of vacuum vessel
+    :type radx_outer: float
+    :param rminx_outer: inboard radius of outer surface of vacuum vessel
+    :type rminx_outer: float
+    :param radx_inner: outboard radius of inner surface of vacuum vessel
+    :type radx_inner: float
+    :param rminx_inner: inboard radius of inner surface of vacuum vessel
+    :type rminx_inner: float
+    :return: radial and vertical coordinates for vacuum vessel geometry below the midplane
+    :rtype: tuple
```
Comment (timothy-nunn) [inline process/geometry/vacuum_vessel_geometry.py:108]: Tuples of what?

File: process/geometry/vacuum_vessel_geometry.py [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+    :param radx_outer: outboard radius of outer surface of vacuum vessel
+    :type radx_outer: float
+    :param rminx_outer: inboard radius of outer surface of vacuum vessel
+    :type rminx_outer: float
+    :param radx_inner: outboard radius of inner surface of vacuum vessel
+    :type radx_inner: float
+    :param rminx_inner: inboard radius of inner surface of vacuum vessel
+    :type rminx_inner: float
+    :return: VacuumVesselGeometry - dataclass returning vacuum vessel radial and vertical coordinates
+    :rtype: DataClass
```
Comment (timothy-nunn) [inline process/geometry/vacuum_vessel_geometry.py:148]: Same comment re return type

**Review by timothy-nunn — CHANGES_REQUESTED** [review]
It looks like a lot of comments, but it basically boils down to:
* Clearing up exactly the types that are contained in the dataclasses and ensuring they have a well-defined API.
* Discussing if consolidation of some of the dataclasses into one unified `ArbitraryGeometry` would make sense (this is not me saying we should do that, it is something for us to discuss)
* Clarifying the names of some functions that are called 'plot...' but dont seem to do any plotting.
* Clarifying the docstrings (types and descriptions) on functions that have lots of returns. This is a complicated sub-package and it will get confusing what does and returns what without more documentation, and I want using this to be as easy as possible for any user of PROCESS.
* Variable naming to, again, aide in clarity and understanding
* Some other minor comments about fudge factors. I see they were in old plot_proc so it might a case of you asking one of the modelers. 

Other than that, the sub-package is laid out well, docstring format is all correct, and plot_proc is looking a lot less busy!

File: process/geometry/plasma_geometry.py [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+        angs1 = np.linspace(
+            -(inang + theta1) + np.pi, (inang + theta1) + np.pi, 256, endpoint=True
+        )
+        angs2 = np.linspace(-(outang + theta2), (outang + theta2), 256, endpoint=True)
+    elif i_single_null < 0:
+        angs1 = np.linspace(
+            -(inang + theta1) + np.pi, theta1 + np.pi, 256, endpoint=True
+        )
+        angs2 = np.linspace(-theta2, (outang + theta2), 256, endpoint=True)
+    else:
```
[the remark this answers is not in the document]
Comment (clmould) [reply process/geometry/plasma_geometry.py:69]: It was like this in the source code

File: process/geometry/plasma_geometry.py [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+    :type triang_95: float
+    :param kappa_95: plasma elongation at 95% surface
+    :type kappa_95: float
+    :param i_single_null: switch for single null / double null plasma
+    :type i_single_null: bool
+    :return: PlasmaGeometry - dataclass returning plasma elongation and radial and vertical coordinates of the plasma
+    :rtype: DataClass
+    """
+    delta = 1.5 * triang_95
+    kappa = (1.1 * kappa_95) + 0.04
```
[the remark this answers is not in the document]
Comment (clmould) [reply process/geometry/plasma_geometry.py:41]: It was like this in the source code

File: process/geometry/plasma_geometry.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (clmould) [reply process/geometry/plasma_geometry.py:69]: I will remove this

**Review by timothy-nunn — APPROVED** [review]
I am happy with the code. Could you maybe comment a before and after from plotproc of the large tokamak just to show nothing has changed?

**Comment by clmould** [comment]
plot_proc output for large tokamak before:
plot_proc_2

after:
image

**Comment by timothy-nunn** [comment]
Thanks Clair, all looks good to me! Thanks for your work on this!

**Pull request merged**